src/rng.jl

```diff
@@ -0,0 +1,31 @@
+"""
+Data structure to keep track of the history of the random stream
+produced by RNG.
+"""
+struct TracedRNG{T} <: Random.AbstractRNG where T <: Random.AbstractRNG
+    count::Base.RefValue{Int}
+    rng::T
+    seed
+end
+
+# Set seed manually, for init ?
+Random.seed!(rng::TracedRNG, seed) = Random.seed!(rng.rng, seed)
+# Reset the rng to the initial seed
+Random.seed!(rng::TracedRNG) = Random.seed!(rng.rng, rng.seed)
+
+TracedRNG() = TracedRNG(Random.seed!()) # Pick up an explicit RNG from Random
+TracedRNG(rng::Random.AbstractRNG) = TracedRNG(Ref(0), rng, rng.seed)
+
+# Dispatch to Random, keep track of numbers generated
+function Random.rand!(rng::TracedRNG, values...)
+    res = Random.rand(rng.rng, values...)
+    n = length(res)
+    inc_count!(rng, n)
+    return res
+end
+
+inc_count!(rng::TracedRNG) = inc_count!(rng, 1)
```

Comment:
To me it still seems that [counter-based PRNGs](https://en.wikipedia.org/wiki/Counter-based_random_number_generator_(CBRNG)) might be better here: they would allow to (re-)generate the n-th random number without having to generate the first n-1 random numbers.

Reply:
I think the implementation is actually in https://github.com/sunoru/Random123.jl.

Reply:
An example on the GPU: https://juliafolds.github.io/FoldsCUDA.jl/dev/examples/monte_carlo_pi/

Reply:
I like the idea, however, I would suggest that we keep this as future work. For this PR, we probably want to pass `rng` around all necessary places first, so that we can fix `forkr` next.

Reply:
:+1: 

I just noticed that DiffEqNoiseProcesses also uses splittable counter-based RNGs: https://github.com/SciML/DiffEqNoiseProcess.jl/blob/master/src/noise_interfaces/virtual_brownian_tree_interface.jl and https://diffeq.sciml.ai/stable/features/noise_process/#VirtualBrownianTree